shared/js/background/redirect.es6.js

```diff
@@ -36,6 +38,40 @@ function buildResponse (url, requestData, tab, isMainFrame) {
     }
 }
 
+function updateTabCleanAmpUrl (currentTab, canonicalUrl, url) {
+    if (currentTab) {
+        currentTab.cleanAmpUrl = canonicalUrl || url
+    }
+}
+
+async function handleAmpAsyncRedirect (thisTab, url) {
+    const canonicalUrl = await ampProtection.fetchAMPURL(thisTab.site, url)
+    const currentTab = tabManager.get({ tabId: thisTab.id })
+    updateTabCleanAmpUrl(currentTab, canonicalUrl, url)
+    if (canonicalUrl) {
+        return { redirectUrl: canonicalUrl }
+    }
+}
+
+async function handleAmpDelayedUpdate (thisTab, url) {
+    const canonicalUrl = await ampProtection.fetchAMPURL(thisTab.site, url)
+    const currentTab = tabManager.get({ tabId: thisTab.id })
+    const newUrl = canonicalUrl || url
+    updateTabCleanAmpUrl(currentTab, canonicalUrl, url)
+
+    browser.tabs.update(thisTab.id, { url: newUrl })
```

Comment:
I think in the `url` case when `canonicalUrl` is null we're going to make an infinite loop no?